Getting rid of Makefile.shared

levitte opened the pull request:
Makefile.shared is a beast that we really should have dealt with for 1.1.0, but there wasn't time.  It's a problematic procedure insofar that it hides away parts of what's actually happening and is very hard to debug.  Better then to move the operation performed to the actual Makefile.  While I'm at it, I've made the use of certain (undocumented) configs a bit more consistent with what's actually needed, and hopefully easier to read.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated

levitte committed:
Configure: Add read_eval_file, a general purpose perl file reader/evaluator

It will return the last expression from the input file.

We also use this in read_config, which slightly changes what's
expected of Configurations/*.conf.  They do not have to assign
%targets specifically.  On the other hand, the table of configs MUST
be the last expression in each of those files.

levitte committed:
Configure: Read in extra information to help create shared libraries

This will replace the use of Makefile.shared

This also means a small adjustment on how the attributes dso_cflags,
dso_cxxflags and dso_lflags are treated.  They were previously treated
as an extension to shared_cflag, shared_cxxflag and shared_ldflag, but
they should really be regarded as alternatives instead, for example
for darwin, where -dynamiclib is used for shared libraries and -bundle
for DSOs.

levitte committed:
Configure: Recognise .rc and .def / .map / .opt as source files

This makes it possible to add build.info statements for using resource
files as well as linker scripts (.def for Windows, .map for Unix, and
.opt for VMS) is if they were source files.  This requires changes in
the build file templates.

levitte committed:
Build file templates: Replace the use of Makefile.shared

Because this also includes handling all sorts of non-object files when
linking a program, shared library or DSO, this also includes allowing
general recognition of files such as .res files (compiled from .rc
files), or .def / .map / .opt files (for export and possibly
versioning of public symbols only).

This does mean that there's a tangible change for all build file
templates: they must now recognise and handle the `.o` extension,
which is used internally to recognise object files internally.  This
extension was removed by common.tmpl before this change, but would
mean that the platform specific templates wouldn't know if "foo.map"
was originally "foo.map.o" (i.e. an object file in its own right) or
"foo.map" (an export definition file that should be treated as such,
not as an object file).

For the sake of simplifying things, we also modify util/mkdef.pl to
produce .def (Windows) and .opt (VMS) files that don't need additional
hackery.

levitte committed:
build.info: adapt to the new handling of .rc / .def / .map / .opt files

levitte committed:
Configure et al: cleanups

Remove some config attributes that just duplicate values that are
already there in other attributes.

Remove the special runs of mkdef.pl and mkrc.pl from build file
templates, as these are now done via GENERATE statements in
build.info.

Remove all references to ordinal files from build file templates, as
these are now treated via the GENERATE statements in build.info.

Also remove -shared flags and similar that are there in shared-info.pl
anyway.  (in the case of darwin, it's mandatory, as -bundle and
-dynamiclib don't mix)

levitte committed:
Remove Makefile.shared, as it's now entirely unused

levitte committed:
Note the removal of Makefile.shared in CHANGES

richsalz commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+    return grep /gcc/, @lines;
+}
+
+my %shared_info;
```
combine the two lines?  And with that, my Perl expertise is all used up :)

levitte replied:
Good point

levitte replied:
(I think...  since I have references to `%shared_info` from within the subs in this structure, I'm worried that combining will mean the symbol isn't lexigraphically defined when those subs are compiled.

levitte replied:
Ahyup, when I combined into one line, I got this:

```
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 32.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 36.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 38.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 54.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 61.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 68.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 75.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 82.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 89.
Global symbol "%shared_info" requires explicit package name (did you forget to declare "my %shared_info"?) at (eval 11) line 96.
```

So, errr, no.

richsalz commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -942,71 +947,77 @@ EOF
+          if defined $target{shared_impflag};
+      my $shared_def = join("", map { ' '.$target{shared_defflag}.$_ } @defs);
+      my $recipe = <<"EOF";
 # With a build on a Windows POSIX layer (Cygwin or Mingw), we know for a fact
```
Since you're in the area, should "With a build on..." be changed to "When building on..." ?  Side-note.

levitte replied:
Will fix

richsalz commented on Configure:
```diff
@@ -882,9 +882,43 @@ my %target = resolve_config($target);
+    }
+}
+use Data::Dumper;
+print STDERR Dumper({ %target });
```
debug leftover?

levitte replied:
Ah, yes, thank you.

richsalz reviewed (approved):
Some nits you can ignore.  approved once the builds pass.  And to be explicit I agree that changing the build system doesn't change our A?I guarantees :)

levitte commented:
Hmmm, I think I need to double check on Windows...  Appveryor seems to hang forever during configuration...

levitte commented:
Hah!  Windows problem resolved, or so I hope.  Please reconfirm (you can just look at the new commit if you prefer...  it will be squashed in appropriately when I'll merge)

richsalz commented:
reconfirm after the builds are done.

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+sub detect_gnu_cc {
+    my @lines =
+        `$config{cross_compile_prefix}$target{cc} -v 2>&1`;
+    return grep /gcc/, @lines;
```
As pointed out in #4755, this is not reliable way to detect gcc.

levitte replied:
Ah...  ok, just gotta ensure %predefined is filled in early enough, then.

levitte replied:
... which isn't so easy, we have a bit of a catch 22 there, since `$target{cc}` is modified far later than these functions are called.  Note that his is copied directly from `Makefile.shared`, and also, this is only called for `svr5-shared`, where I'm unsure there's been any report of an issue (not that we get much reports from that platform anyway).  So I'd like to argue that we can leave this alone for this PR and solve in a more general in a separate effort (I know you're thinking of options).  Sounds reasonable to you?

dot-asm replied:
Here is a wild idea... Why can't those flags that are defined here in shared-info.pl be just a template? I mean why can't a target be described as `inherit_from => [ ..., "something-shared", ... ]`? Why is there additional level of "indirection" with shared_target "pointing" to separate data structure? Wouldn't it be sufficient if shared_target was rather just a "reference" to a *subroutine* that accepts complete target description as argument? And not dedicated data structure and target that is?

levitte replied:
'cause inherit_from is processed before the final target structure is done, and because the subs that are values for attributes only receive the inherited value for that same attribute. In the other PR, where I implement later processing of those subs, it may be possible to get any target value from any sub, that will make things easier...

levitte replied:
Forgot to say that some of those shared-info.pl structures depend on $target{cc}, so you can see how that's a problem to handle with inherit_from

dot-asm replied:
And the question is does it *have to* be that way? One can also put it in following way. This is kind of dominated by idea that it *has to* mimic Makefile.shared, but does it actually have to?

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+    'bsd-shared' => sub {
+        return $shared_info{'gnu-shared'} if detect_gnu_ld();
+        return {
+            shared_ldflag     => '-shared -nostlib',
```
-nostdlib, extra 'd'.

levitte replied:
Fixed

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+        };
+    },
+    'solaris-shared' => sub {
+        return $shared_info{'gnu-shared'} if detect_gnu_ld();
```
This can be removed, because there is no gnu ld on Solaris, and even gcc uses vendor linker.

levitte replied:
Ok...  this came from early experimentation on Solaris around 2000, and I don't quite recall on what platform I saw the need for this.  I guess I got paranoid and did this for all platforms where we used gcc at all.

Will ix.

levitte replied:
Fixed, will push soon

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+            shared_ldflag     => '-G',
+            shared_sonameflag => '-h ',
+        };
+    },
```
One can wonder what are the options of having this in .conf files. I mean this SCO stuff doesn't really belong in 10-main.conf. So if there are any nostalgic feelings it, SCO stuff should rather be moved to 80-sco.conf, and it such case it would be only appropriate if this is also moved. So that it can be later deleted in one step...

levitte replied:
I'm actually thinking of making config templates instead of this special case file, or where appropriate, just insert these values in already existing templates.  We could also get rid of the `shared_target` attribute and add a `enable => [ 'shared' ]` instead (since we already have the processing of such attributes in place).

levitte replied:
Ah, sorry, this will actually be difficult to do.  The reason is that the check of GNU ld in some of the shared targets would need to happen in `inherit_from`, which is computed early ('cause reasons), even in #4854.  `$target{cc}` isn't defined then, getting us back in a catch 22 that is even harder to resolve than the one I mentioned [above](https://github.com/openssl/openssl/pull/4840#discussion_r154786498)...
(I think it's still possible, but will be an uglier hack...)

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,108 @@
+        };
+    },
+    'hpux-shared' => sub {
+        return $shared_info{'gnu-shared'} if detect_gnu_ld();
```
I'm pretty sure there is no gnu ld on HP-UX either...

dot-asm replied:
"Pretty sure" means that I don't see one on 11 and have hard time imagining that there was one earlier.

levitte replied:
Okie, removed

dot-asm commented on build.info:
```diff
@@ -26,6 +24,52 @@ GENERATE[crypto/include/internal/bn_conf.h]=crypto/include/internal/bn_conf.h.in
+  SHARED_SOURCE[libssl]=libssl.symvec
+ENDIF
+
+IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-)/ -}]
```
Wouldn't `VC\-` be more appropriate? I mean since dash has special meaning. Well, it does work now...

levitte replied:
The dash is only special within `[]`, so no.

glensc replied:
and escaping `-` in `[]` is **not** done with `\` (doing so would include literal `\`), but rather having it as last or first char inside `[]`.

levitte commented:
Travis has a failure in `70-test_tls13kexmodes.t` that I cannot relate to this change...  and can't reproduce either.

levitte commented:
@dot-asm, you might be interested in looking at a branch where I have merged together this PR, #4854 and #4899, [found here](https://github.com/levitte/openssl/tree/getting-rid-of-Makefile.shared%2BConfigure-rething-target-resolving-2%2BConfigure-predefined-function), to show how it all comes together.

levitte commented:
If anything, this will make those who appreciate the silence of `make -s` happy ;-)

richsalz commented:
There's a reason why I approved this :)  And reconfirm now.

levitte commented:
I hear ya...  but I'm gonna wait for @dot-asm to say his piece as well...  and I just realised that there are things I haven't fixed that I said I would, so ;-)

dot-asm commented:
Red cross from appveyor ought to be relevant.

dot-asm commented:
I get failures on MacOS X. Quite old though. Could you temporarily throw osx back into .travis.yml?

levitte commented:
As you can see from the fixup / squash commits I'm throwing in, I'm currently watching all CIs quite closely ;-)

Sure, I'll trow in osx as well (should we make that a permanent thing?)

levitte committed:
TEMPORARY: Add osx back for Travis

clang on osx seems to currently fail on '%jd' specs, so turn off the check
of printf format there.

dot-asm commented:
> should we make that [osx] a permanent thing?

It was taken away because it was common queue, and as results of that request turn-around was effectively bound to backlog on osx, which was huge [during working days]. And still is according to https://www.traviscistatus.com/#week, so it's useless for general use...

levitte commented:
Ah, yeah, that doesn't look too good

levitte commented:
Seeing some preliminary osx results...  it gets stuck at `%jd`.  Meh...

dot-asm commented:
I get errors linking shared library! I suppose I have to get my hands on newer mac...

dot-asm commented:
BTW, the error I get is `ld: malformed 64-bit a.b.c.d.e version number: -compatibility_version`. It's on MacOS X something-not-supported-anymore. Because it's running on Mac that shouldn't exits, so that OS can't be upgraded. It's originally 32-bit mini with 64-bit CPU taken from broken Dell laptop...

levitte commented:
Oh interesting....  well, I've just hacked .travis.yml so it shouldn't bail on `%jd`, let's see what we get from that

dot-asm commented:
Well, same error message *is* reported in *first* failed build, e.g. https://travis-ci.org/openssl/openssl/jobs/314947419, so it's not specific to too-old-to-be-true-version.

levitte commented:
Ah, easily fixed

dot-asm commented:
`clang: error: invalid argument '-bundle' not allowed with '-dynamiclib'` when linking a shared engine.

levitte commented:
Hmmmm...  I'll have to check that out, they aren't supposed to come together...

dot-asm commented:
Just in case, I'm signing off for today soon...

levitte commented:
Something I'm wonder about...  there are a number of '*-gcc' targets in `Configurations/10-main.conf` that have `shared_ldflag => '-shared'` (sometimes a little more, but at least that)...  but their `shared_target` indicate an `ld` that uses other flags, and I can see their `*-cc` counterparts using those flags.  So I wonder what gives?  After all, `-shared` is primarly a GNU ld thing, isn't it?

levitte commented:
In the mean time, I found the error with `-bundle` and `-dynamiclib`...  more cleaning out in `Configurations/10-main.conf`

dot-asm commented:
> -shared is primarly a GNU ld thing, isn't it?

Not quite. It's rather a gcc driver thing, i.e. one that calls different components, including ld. gcc driver is supposed to know how to translate -shared to whatever suitable under circumstances. For example on Solaris gcc calls vendor ld, and so -shared is supposed to pass vendor-specific flags, *minimally* required.

levitte commented:
Ah, ok.  Thanks

levitte commented:
The remaining osx issue is unrelated to this PR

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -865,7 +862,7 @@ sub vms_info {
         dso_scheme       => "dlfcn",
         shared_target    => "linux-shared",
         shared_cflag     => "--pic",
-        shared_ldflag    => add("-z --sysv --shared"),
```
I'd like to insist on keeping this. Note two dashes, i.e. it's not something that is common. However, this config is not actually fully operational [yet], so if one has to actually do something with it *now*, it probably would be more appropriate to say 'shared_target => "unknown" or remove 'shared_target' altogether?

levitte replied:
Makes sense, I was apprehensive of that one.  I'll remove `shared_target`

levitte replied:
I decided the other way, to leave it as it was (i.e. --shared restored).  I don't know the platform, so less sudden surprises, right?

dot-asm replied:
I know everything about it, and I find it hard to imagine that anybody else knows something. So either way is fine... I mean as long as --shared stays...

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -311,9 +310,7 @@ sub vms_info {
         cflags           => add_before("-xarch=v9"),
-        lflags           => add_before("-xarch=v9"),
         bn_ops           => "BN_LLONG RC4_CHAR",
-        shared_ldflag    => "-xarch=v9 -G -dy -z text",
```
So suggestion is to remove this, but not corresponding line in solaris64-x86_64-cc... Which one is right?

levitte replied:
Well, you already have `-xarch=v9` in cflags, and `-G -dy -z text` gets inherited from `solaris-sparcv7-cc`, so having them here is just redundant.  `solaris64-x86_64-cc` doesn't inherit its flags the same way, although it's true that `-xarch=generic64` is redundant there...  is that what you meant?

I didn't go in a all-out-clean-everything-meticulously, there were just a few things I happened to see that I cleaned out.  Are you asking me to be more thorough here?

dot-asm replied:
Ah! I see now... Keyword is that solaris64-x86_64-cc doesn't inherit. Thanks. As for -xarch=generic64 being redundant there. I didn't pay attention to *that*, but it *is* redundant. For consistency it could and probably should be removed.

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1594,29 +1588,29 @@ sub vms_info {
         cflags           => add("-arch ppc -std=gnu9x -DB_ENDIAN -Wa,-force_cpusubtype_ALL"),
         perlasm_scheme   => "osx32",
-        shared_ldflag    => "-arch ppc -dynamiclib",
+        shared_ldflag    => "-arch ppc",
```
Originally shared libraries were linked without `$(CFLAGS)`. And this was the reason for why one had to have things like -arch/-xarch/-m64 in ldflags. Now they are linked with `$(CFLAGS)` that already has -arch/-xarch/-m64 and so defining them in ldflags is actually redundant. In other words if -dynamic link is no longer needed, then one can as well just remove this line. As well as corresponding lines in all `darwin*` targets.

levitte replied:
Good point

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,102 @@
+        shared_sonameflag     => '-install_name $(INSTALLTOP)/$(LIBDIR)/',
+    },
+    'cygwin-shared' => {
+        shared_ldflag         => '-shared -Wl,--enable-auto-image-base -Wl,-Bsymbolic',
```
On side[!] note. I don't think that -Bsymbolic makes any difference on Windows, be it cygwin, mingw, or whatever. Because *executable format* doesn't have provisions for non-Bsymbolic kind of linking.

levitte replied:
Ok sure.  It's just that it was there in Makefile.shared.  Are you saying it should be safe to remove?

dot-asm replied:
Do note that it was a *side* note. But I *am* saying that it *should be safe* to remove. However, there is difference between "should be safe to remove" and "should be removed [now]". One can still argue that it can be removed now, because we will forget to look into it later...

levitte replied:
I know it's a side note, but I'm thinking that while we're doing some *light* cleanup in the process, might as well.  Done.

dot-asm commented on Configurations/shared-info.pl:
```diff
@@ -0,0 +1,102 @@
+    },
+    'cygwin-shared' => {
+        shared_ldflag         => '-shared -Wl,--enable-auto-image-base -Wl,-Bsymbolic',
+        shared_impflag        => '-Wl,--out-implib=',
```
On side[!] note. I'm not actually fond of this kind of things. For following reason. There is a code somewhere that takes care of it. It ought to check if this variable is defined or not and act accordingly. It also ought to make an *assumption* that variable is defined in manner that permits appending a name. In other words complex logic nowhere near variable assignment. I would find it more appropriate if it was simply added to shared_ldflags with some kind of indirect reference to basename of currently processed library. Yes, it would mean that there would have to be code somewhere that processed the indirect reference, but that could be much more generic than checking for specific variable...

levitte replied:
Not sure how you figure the indirect reference to look like.  You can't use make variables, so that would be either to insert som perl code in there (I fear that would result in eval hell), or template stuff (perl within `{-` and `-}`, which leads to some kind of eval as well, but I actually find that less scary), or adding a sort of mini language (such as having `-Wl,--out-implib={LIB}` in `shared_ldflag`).
So while I value the idea (and "side[!] note" *is* noted), I think that's going to have to be for another PR, 'cause it needs some more thinking not to cause a larger mess.

dot-asm replied:
> that's going to have to be for another PR,

If *ever*! I mean I'm only saying I'm not *fond of* it and it might be worth considering something that would be more generalized :-)

levitte replied:
I understand that, and I actually agree...  thing is, something more generalized might end up more magic and with a load of complexity under the hood.  That's why I'm calling for caution.  But we should talk more about it (perhaps not right here, though, so we don't side track this PR too much)

dot-asm commented:
I'll take it for a spin on HP-UX and AIX under the day...

levitte commented:
I did a squash spree...  I hope that's alright

levitte committed:
fixup! Configure et al: cleanups

levitte committed:
squash! Configure: Read in extra information to help create shared libraries

Added note:

We take the opportunity to clean out things that are redundant or
otherwise superfluous (for example the check of GNU ld on platforms
where it never existed).

dot-asm commented:
I get a lot of linker warnings for aix-gcc, but make test passes... I haven't looked yet if it's specific to this request, or warnings were there all along... Checking aix-cc...

dot-asm commented:
Hmm... No warnings with aix-cc...

dot-asm commented:
aix-gcc warnings are not specific to this request, so that if they are to be fixed, it should be done separately...

dot-asm commented:
HP-UX work. [On side note, I'm not convinced that removal of equivalent of automatic LIBRPATH is justified *universally*. I mean it was removed, because it made some sense on Linux, but the argument was extended to all platforms, probably unjustified. So I'm likely to revisit it at some point...]

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -243,12 +243,12 @@ sub vms_info {
                                        threads("-D_REENTRANT")),
         thread_scheme    => "pthreads",
-        lflags           => add("-xarch=generic64",threads("-mt")),
+        lflags           => add(threads("-mt")),
```
Oh! So in other words lflags are also always used with cflags. Fair enough...

levitte replied:
Yup. If you look at Makefile.shared before this change, you will see that CFLAGS and LDFLAGS were always used when linking apps. Those are populated from cflags and lflags respectively (sometim

levitte commented:
Merged.  3b6c4b07364797566c2c1fd75e499b2d9dd73506 to 5f0e171a10325ec4502c2ce41b56d46f3c121fcb

levitte commented:
Thank you for the careful review
